components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/DefaultOIDCClaimsCallbackHandler.java

```diff
@@ -604,6 +617,31 @@ private Map<String, Object> getUserClaimsInOIDCDialect(String spTenantDomain,
             }
         }
 
+        /*
+        If the application requested for groups and a shared user is accessing a shared org of that user,
+        get the groups of the shared user from the shared organization.
+        */
+        if (requestedClaimUris.contains(GROUPS_CLAIM) && isSharedUserAccessingSharedOrg(authenticatedUser) &&
+                StringUtils.isNotEmpty(authenticatedUser.getSharedUserId())) {
+            List<String> requestedClaimForSharedUser = new ArrayList<>();
+            requestedClaimForSharedUser.add(GROUPS_CLAIM);
+            // Getting the accessing tenant domain to get the userstore manager of the shared organization.
+            userAccessingTenantDomain = OAuthComponentServiceHolder.getInstance().getOrganizationManager()
+                    .resolveTenantDomain(authenticatedUser.getAccessingOrganization());
+            AbstractUserStoreManager userStoreManager =
+                    (AbstractUserStoreManager) OAuthComponentServiceHolder.getInstance().getRealmService()
+                            .getTenantUserRealm(IdentityTenantUtil.getTenantId(userAccessingTenantDomain))
+                            .getUserStoreManager();
+            String fullQualifiedSharedUsername = userStoreManager.getUser(authenticatedUser.getSharedUserId(), null)
+                    .getFullQualifiedUsername();
+            UserRealm sharedUserRealm = IdentityTenantUtil.getRealm(userAccessingTenantDomain,
+                    fullQualifiedSharedUsername);
+            // Getting the shared user's group claim from the shared organization.
+            Map<String, String> sharedUserGroupClaim = getUserClaimsInLocalDialect(fullQualifiedSharedUsername,
+                    sharedUserRealm, requestedClaimForSharedUser);
+            userClaims.put(GROUPS_CLAIM, sharedUserGroupClaim.get(GROUPS_CLAIM));
+        }
```

Comment:
shall we extract this to a private method?

Reply:
Done